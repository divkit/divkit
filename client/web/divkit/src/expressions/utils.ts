import type { EvalContext, EvalValue, EvalValueBase } from './eval';
import type { Node, Variable } from './ast';
import type { EvalTypes } from './eval';
import type { VariablesMap } from './eval';
import { walk } from './walk';
import { parseColor, ParsedColor } from '../utils/correctColor';
import { padLeft } from '../utils/padLeft';
import { hasBigInt, MAX_INT, MIN_INT, toBigInt } from './bigint';

export function valToInternal(val: EvalValue): EvalValue {
    if (val.type === 'url' || val.type === 'color') {
        return {
            type: 'string',
            value: val.value
        };
    }

    return val;
}

export function dateToString(date: Date): string {
    return date
        .toISOString()
        .replace('T', ' ')
        .replace(/\.\d{3}Z$/, '');
}

export function valToString(val: EvalValue): string {
    if (val.type === 'string') {
        return val.value;
    } else if (val.type === 'integer') {
        return String(val.value);
    } else if (val.type === 'number') {
        let res = String(val.value);

        if (!res.includes('.')) {
            if (res.includes('e')) {
                res = res.replace('e', '.0e');
            } else {
                res += '.0';
            }
        }

        res = res.replace(/e\+?/i, 'E');

        return res;
    } else if (val.type === 'boolean') {
        return val.value ? 'true' : 'false';
    } else if (val.type === 'datetime') {
        return dateToString(val.value);
    } else if (val.type === 'color') {
        return stringifyColor(safeConvertColor(val.value));
    } else if (val.type === 'url') {
        return val.value;
    } else if (val.type === 'dict') {
        return '<dict>';
    }

    // For purpose when new eval value types will be added
    throw new Error(`Unexpected type ${(val as EvalValueBase).type}`);
}

export function valToPreview(val: EvalValue): string {
    let res = valToString(val);

    if (val.type === 'string') {
        res = "'" +
            res
                .replace(/\\/g, '\\\\')
                .replace(/'/g, '\\\'') +
            "'";
    }

    return res;
}

export function typeToString(type: EvalTypes): string {
    if (type === 'datetime') {
        return 'DateTime';
    }
    return type.charAt(0).toUpperCase() + type.substring(1);
}

export function roundInteger(ctx: EvalContext, val: number | bigint): number | bigint {
    if (hasBigInt) {
        return toBigInt(val);
    }

    if (val < 0) {
        const res = -Math.floor(-val as number);
        if (res === 0) {
            // -0 => 0
            return 0;
        }
        return res;
    }

    return Math.floor(val as number);
}

export function checkIntegerOverflow(ctx: EvalContext, val: number | bigint): void {
    if (val < MIN_INT || val > MAX_INT) {
        throw new Error('Integer overflow.');
    }

    if (!hasBigInt && (val > Number.MAX_SAFE_INTEGER || val < Number.MIN_SAFE_INTEGER)) {
        ctx.safeIntegerOverflow = true;
    }
}

export function gatherVarsFromAst(ast: Node): string[] {
    const res: string[] = [];

    walk(ast, {
        Variable(node: Variable): void {
            res.push(node.id.name);
        }
    });

    return res;
}

export function evalError(msg: string, details: string): never {
    throw new Error(`Failed to evaluate [${msg}]. ${details}`);
}

export function containsUnsetVariables(ast: Node, variables: VariablesMap): boolean {
    let result = false;

    walk(ast, {
        Variable(node: Variable): void {
            if (!variables.has(node.id.name)) {
                result = true;
            }
        }
    });

    return result;
}

export function safeConvertColor(color: string): ParsedColor {
    const res = parseColor(color);

    if (res) {
        return res;
    }

    throw new Error('Unable to convert value to Color, expected format #AARRGGBB.');
}

export function stringifyColor(color: ParsedColor): string {
    return `#${[color.a, color.r, color.g, color.b].map(it => {
        if (it < 0 || it > 255) {
            throw new Error('Value out of range 0..1.');
        }

        return padLeft(Math.round(it).toString(16), 2);
    }).join('').toUpperCase()}`;
}

export function transformColorValue(color: string): string {
    return stringifyColor(safeConvertColor(color));
}
