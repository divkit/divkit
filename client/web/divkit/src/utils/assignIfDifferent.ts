import { isDeepEqual } from './isDeepEqual';

export function assignIfDifferent<A, B>(newVal: A, defaultVal: B): A | B {
    if (isDeepEqual(newVal, defaultVal)) {
        return defaultVal;
    }
    return newVal;
}
