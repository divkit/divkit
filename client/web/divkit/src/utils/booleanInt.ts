import type { BooleanInt } from '../../typings/common';

export function booleanInt(val: BooleanInt): boolean {
    return val === true || val === 1;
}
