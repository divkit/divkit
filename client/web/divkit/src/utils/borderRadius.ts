import type { CornersRadius } from '../types/border';
import { pxToEm } from './pxToEm';

export function borderRadius(cornersRadius: CornersRadius, defaultRadius = 0, fontSize = 10): string {
    return [
        cornersRadius['top-left'],
        cornersRadius['top-right'],
        cornersRadius['bottom-right'],
        cornersRadius['bottom-left']
    ]
        .map(it => pxToEm((it || defaultRadius) / fontSize * 10))
        .join(' ');
}
