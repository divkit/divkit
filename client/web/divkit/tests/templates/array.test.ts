import {
    describe,
    expect,
    test,
    vi
} from 'vitest';

import { applyTemplate } from '../../src/utils/applyTemplate';

describe('array', () => {
    test('test_array_empty', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_empty.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_invalid_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_invalid_items.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_invalid_items_in_data', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_invalid_items_in_data.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_missing_data', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_missing_data.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_missing_one_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_missing_one_link.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_nested', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_nested.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_nested_with_internal_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_nested_with_internal_link.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_not_templated', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_not_templated.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_one_invalid_item', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_one_invalid_item.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_complex_items_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_complex_items_with_internal_links.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_heterogeneous_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_heterogeneous_items.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_heterogeneous_items_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_heterogeneous_items_with_internal_links.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_internal_links.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_link.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_nested_templated_array', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_nested_templated_array.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_override_in_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_override_in_items.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_override_in_template', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_override_in_template.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_templated_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_templated_items.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_without_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_without_link.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });
});
