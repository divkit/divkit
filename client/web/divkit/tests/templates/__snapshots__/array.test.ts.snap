// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array > test_array_empty 1`] = `
{
  "json": {
    "array": [],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array > test_array_invalid_items 1`] = `
{
  "json": {
    "array": [
      {},
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array > test_array_invalid_items_in_data 1`] = `
{
  "json": {
    "array": [
      {},
    ],
    "link": [
      {},
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {},
    ],
  },
}
`;

exports[`array > test_array_missing_data 1`] = `
{
  "json": {
    "array": [
      {
        "type": "entity_with_complex_property",
      },
      {
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array > test_array_missing_data 2`] = `[]`;

exports[`array > test_array_missing_one_link 1`] = `
{
  "json": {
    "array": [
      {
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array > test_array_missing_one_link 2`] = `[]`;

exports[`array > test_array_nested 1`] = `
{
  "json": {
    "array": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "link": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
  },
}
`;

exports[`array > test_array_nested_with_internal_link 1`] = `
{
  "json": {
    "array": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "nested_array_link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "nested_array_link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array > test_array_not_templated 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array > test_array_not_templated 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array > test_array_one_invalid_item 1`] = `
{
  "json": {
    "array": [
      {},
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array > test_array_with_complex_items_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": "https://ya.ru",
    "link_2": "https://ya.ru",
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": "https://ya.ru",
    "link_2": "https://ya.ru",
  },
}
`;

exports[`array > test_array_with_heterogeneous_items 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
  },
}
`;

exports[`array > test_array_with_heterogeneous_items_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "c_link": "https://ya.ru",
    "r_link": "Some text",
    "r_r_link": "second",
    "type": "entity_with_array",
  },
  "templateContext": {
    "c_link": "https://ya.ru",
    "r_link": "Some text",
    "r_r_link": "second",
  },
}
`;

exports[`array > test_array_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array > test_array_with_link 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array > test_array_with_nested_templated_array 1`] = `
{
  "json": {
    "array": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
    "tabs_items_link": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "tabs_items_link": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
  },
}
`;

exports[`array > test_array_with_override_in_items 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "c_0",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "c_0",
      },
    ],
    "link_override": "https://ya.ru",
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_override": "https://ya.ru",
  },
}
`;

exports[`array > test_array_with_override_in_template 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_override": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "t_0",
  },
  "templateContext": {
    "link_override": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array > test_array_with_templated_items 1`] = `
{
  "json": {
    "array": [
      {
        "link": "https://ya.ru",
        "type": "c_0",
      },
      {
        "link": "https://ya.ru",
        "type": "c_0",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array > test_array_with_templated_items 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array > test_array_without_link 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_empty 1`] = `
{
  "json": {
    "array": [],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_invalid_items 1`] = `
{
  "json": {
    "array": [
      {},
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_invalid_items_in_data 1`] = `
{
  "json": {
    "array": [
      {},
    ],
    "link": [
      {},
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {},
    ],
  },
}
`;

exports[`array test_array_missing_data 1`] = `
{
  "json": {
    "array": [
      {
        "type": "entity_with_complex_property",
      },
      {
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_missing_data 2`] = `[]`;

exports[`array test_array_missing_one_link 1`] = `
{
  "json": {
    "array": [
      {
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array test_array_missing_one_link 2`] = `[]`;

exports[`array test_array_nested 1`] = `
{
  "json": {
    "array": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "link": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
  },
}
`;

exports[`array test_array_nested_with_internal_link 1`] = `
{
  "json": {
    "array": [
      {
        "array": [
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
          {
            "property": {
              "value": "https://ya.ru",
            },
            "type": "entity_with_complex_property",
          },
        ],
        "type": "entity_with_array",
      },
    ],
    "nested_array_link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "nested_array_link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array test_array_not_templated 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_not_templated 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array test_array_one_invalid_item 1`] = `
{
  "json": {
    "array": [
      {},
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array test_array_with_complex_items_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": "https://ya.ru",
    "link_2": "https://ya.ru",
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": "https://ya.ru",
    "link_2": "https://ya.ru",
  },
}
`;

exports[`array test_array_with_heterogeneous_items 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
  },
}
`;

exports[`array test_array_with_heterogeneous_items_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": "Some text",
        "type": "entity_with_required_property",
      },
      {
        "property": "second",
        "type": "entity_with_string_enum_property",
      },
    ],
    "c_link": "https://ya.ru",
    "r_link": "Some text",
    "r_r_link": "second",
    "type": "entity_with_array",
  },
  "templateContext": {
    "c_link": "https://ya.ru",
    "r_link": "Some text",
    "r_r_link": "second",
  },
}
`;

exports[`array test_array_with_internal_links 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_1": {
      "value": "https://ya.ru",
    },
    "link_2": {
      "value": "https://ya.ru",
    },
  },
}
`;

exports[`array test_array_with_link 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "link": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array test_array_with_nested_templated_array 1`] = `
{
  "json": {
    "array": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
    "tabs_items_link": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {
    "tabs_items_link": [
      {
        "gallery_items_link": [
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
          {
            "item_link": "https://ya.ru",
            "type": "gallery_item",
          },
        ],
        "type": "gallery",
      },
    ],
  },
}
`;

exports[`array test_array_with_override_in_items 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "c_0",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "c_0",
      },
    ],
    "link_override": "https://ya.ru",
    "type": "entity_with_array",
  },
  "templateContext": {
    "link_override": "https://ya.ru",
  },
}
`;

exports[`array test_array_with_override_in_template 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link_override": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "t_0",
  },
  "templateContext": {
    "link_override": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`array test_array_with_templated_items 1`] = `
{
  "json": {
    "array": [
      {
        "link": "https://ya.ru",
        "type": "c_0",
      },
      {
        "link": "https://ya.ru",
        "type": "c_0",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;

exports[`array test_array_with_templated_items 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array test_array_without_link 1`] = `
{
  "json": {
    "array": [
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      {
        "property": {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_array",
  },
  "templateContext": {},
}
`;
